core/src/main/scala/org/http4s/Uri.scala

```diff
@@ -1192,39 +1186,45 @@ object Uri {
         option1.backtrack.orElse(option2)
       }
 
-      (repN(6, h16Colon).with1 ~ ls32)
-        .map { case (ls: collection.Seq[Short], rs) => toIpv6(ls, rs) }
+      (h16Colon.repExactlyAs[List[Short]](6) ~ ls32)
+        .map { case (ls: List[Short], rs) => toIpv6(ls.toList, rs) }
         .backtrack
-        .orElse((doubleColon *> repN(4, h16Colon, P.unit) ~ ls32)
+        .orElse(
+          (doubleColon *> h16Colon.repExactlyAs[List[Short]](5) ~ ls32)
           .map { case (rs: List[Short], rs2) => toIpv6(Seq.empty, rs ++ rs2) })
         .backtrack
-        .orElse(((h16.?.with1 <* doubleColon) ~ h16Colon.repExactlyAs[List[Short]](3).backtrack ~ ls32)
+        .orElse(((h16.?.with1 <* doubleColon) ~ h16Colon.repExactlyAs[List[Short]](4).backtrack ~ ls32)
           .map { case ((ls: Option[Short], rs), rs2) => toIpv6(ls.toSeq, rs ++ rs2) })
         .backtrack
-        .orElse(
-          ((repN(2, h16, colon.void).?.with1 <* doubleColon) ~ h16Colon.repExactlyAs[List[Short]](2).backtrack ~ ls32)
-            .map { case ((ls: Option[List[Short]], rs), rs2) =>
-              toIpv6(ls.getOrElse(Seq.empty), rs ++ rs2)
+        .orElse(((h16.repSep0(0, 2, colon).with1 <* doubleColon) ~ h16Colon.repExactlyAs[List[Short]](3).backtrack ~ ls32)
+            .map { case ((ls: List[Short], rs), rs2) =>
+              toIpv6(ls, rs ++ rs2)
             })
         .backtrack
-        .orElse(((repN(3, h16, colon.void).?.with1 <* doubleColon) ~ h16Colon ~ ls32)
-          .map { case ((ls: Option[List[Short]], r0: Short), rs) =>
-            toIpv6(ls.getOrElse(Seq.empty), Seq(r0) ++ rs)
-          })
+        .orElse(((h16.repSep0(0, 3, colon).with1 <* doubleColon) ~ h16Colon.repExactlyAs[List[Short]](2).backtrack ~ ls32)
```

Comment:
Why (0, 3) instead of (3, 3)?

Reply:
it would fail on e.g. `53f6::1:551b:7bd8`

the number of segments after `::` are always fixed to a number n. so before the `::` we can parse up to 8 - n segments

Reply:
Ah, right. So this is part of the bug fix.